Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -220,6 +247,65 @@ class SerializerTest : public ::testing::Test {
     EXPECT_EQ(type, actual_model.type());
     EXPECT_EQ(model, actual_model);
   }
+
+  void ExpectSerializationRoundTrip(
+      const DocumentKey& key,
+      const FieldValue& value,
+      const SnapshotVersion& update_time,
+      const google::firestore::v1beta1::BatchGetDocumentsResponse& proto) {
+    std::vector<uint8_t> bytes = EncodeDocument(&serializer, key, value);
+    google::firestore::v1beta1::Document actual_proto;
+    bool ok = actual_proto.ParseFromArray(bytes.data(), bytes.size());
+    EXPECT_TRUE(ok);
+
+    // TODO(rsgowman): Right now, we only support Document (and don't support
+    // NoDocument). That should change in the next PR or so.
+    EXPECT_TRUE(proto.has_found());
+
+    // Slight weirdness: When we *encode* a document for sending it to the
+    // backend, we don't encode the update_time (or create_time). But when we
+    // *decode* a document, we *do* decode the update_time (though we still
+    // ignore the create_time). Therefore, we'll verify the update_time
+    // independently, and then strip it out before comparing the rest.
+    EXPECT_FALSE(actual_proto.has_create_time());
+    EXPECT_EQ(update_time.timestamp().seconds(),
+              proto.found().update_time().seconds());
+    EXPECT_EQ(update_time.timestamp().nanoseconds(),
+              proto.found().update_time().nanos());
+    google::firestore::v1beta1::BatchGetDocumentsResponse proto_copy{proto};
+    proto_copy.mutable_found()->clear_update_time();
+    proto_copy.mutable_found()->clear_create_time();
+
+    EXPECT_TRUE(msg_diff.Compare(proto_copy.found(), actual_proto))
+        << message_differences;
+  }
+
+  void ExpectDeserializationRoundTrip(
+      const DocumentKey& key,
+      const FieldValue& value,
+      const SnapshotVersion& update_time,
+      const google::firestore::v1beta1::BatchGetDocumentsResponse& proto) {
+    size_t size = proto.ByteSizeLong();
+    std::vector<uint8_t> bytes(size);
+    bool status = proto.SerializeToArray(bytes.data(), size);
+    EXPECT_TRUE(status);
+    StatusOr<std::unique_ptr<MaybeDocument>> actual_model_status =
+        serializer.DecodeMaybeDocument(bytes);
+    EXPECT_OK(actual_model_status);
+    std::unique_ptr<MaybeDocument> actual_model =
+        std::move(actual_model_status).ValueOrDie();
+
+    // TODO(rsgowman): Right now, we only support Document (and don't support
+    // NoDocument). That should change in the next PR or so.
+    EXPECT_EQ(MaybeDocument::Type::Document, actual_model->type());
+    Document* actual_doc_model = dynamic_cast<Document*>(actual_model.get());
```

Comment:
Will this fail on platforms with no RTTI? Note that it's possible to downcast using `static_cast` (cast to the wrong type is undefined behavior, but the check from the previous line should ensure the type is correct).

Reply:
Whoops; I meant to remove that. Fixed.

(Yeah, that would cause a compilation error with -fno-rtti)